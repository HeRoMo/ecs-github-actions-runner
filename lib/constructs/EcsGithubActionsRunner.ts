import { AutoScalingGroup, TerminationPolicy } from 'aws-cdk-lib/aws-autoscaling';
import {
  Peer, Port, SecurityGroup, Vpc,
} from 'aws-cdk-lib/aws-ec2';
import {
  AsgCapacityProvider,
  Cluster,
  ClusterProps,
  ContainerImage,
  Ec2TaskDefinition,
  LogDriver,
} from 'aws-cdk-lib/aws-ecs';
import { LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';

import { CONFIG, NodeConfig } from '../Config';

/**
 * Information of Github actions runner cluster.
 */
export interface EcsGithubActionsRunnerInfo {
  clusterName: string;
  containerName: string;
  taskDefFamily: string;
  capacityProviders: string[];
}

/**
 * Constructer of ECS cluster for Github actions runner
 */
export class EcsGithubActionsRunner extends Construct {
  /**
   * ECS Cluster
   */
  public readonly cluster: Cluster;

  /**
   * Security Group of Cluster node
   */
  private readonly securityGroup: SecurityGroup;

  /**
   * Task definition family name
   */
  private readonly taskDefinitionFamily: string;

  /**
   * Container name of Github actions runnter.
   */
  private readonly runnerContainerName: string;

  /**
   * Capacity Provider names of Github actions runner cluster.
   */
  private readonly capacityProviders: string[] = [];

  /**
   * Constructor.
   *
   * @param scope Scope
   * @param clusterName Cluster name
   * @param vpc VPC instance
   */
  constructor(
    scope: Construct,
    private readonly clusterName: string,
    private readonly vpc: Vpc,
  ) {
    super(scope, clusterName);

    const clusterProps: ClusterProps = {
      clusterName: this.clusterName,
      vpc: this.vpc,
    };

    this.taskDefinitionFamily = this.clusterName;
    this.runnerContainerName = 'github-actions-runner';

    this.cluster = new Cluster(this, this.clusterName, clusterProps);
    this.securityGroup = this.createSecurityGroup();
    Object.entries(CONFIG.nodes).forEach(([key, node]) => {
      const provider = this.addAsgCapacity(key, node);
      this.capacityProviders.push(provider.capacityProviderName);
    });

    this.createTaskDefinition();
  }

  /**
   * get Cluster infomation object.
   */
  public get clusterInfo(): EcsGithubActionsRunnerInfo {
    return {
      clusterName: this.clusterName,
      containerName: this.runnerContainerName,
      taskDefFamily: this.taskDefinitionFamily,
      capacityProviders: this.capacityProviders,
    };
  }

  /**
   * Create Task Definition
   */
  private createTaskDefinition() {
    const logGroup = new LogGroup(this, `${this.clusterName}-lg`, {
      logGroupName: `/ecs/${this.clusterName}`,
      retention: RetentionDays.ONE_WEEK,
    });

    const taskDefinition = new Ec2TaskDefinition(this, `${this.clusterName}-td`, {
      family: this.taskDefinitionFamily,
    });

    const container = taskDefinition.addContainer('github-actions-runner', {
      containerName: this.runnerContainerName,
      memoryLimitMiB: CONFIG.memoryLimitMiB,
      hostname: this.clusterName,
      image: ContainerImage.fromRegistry('ghcr.io/heromo/ecs-github-actions-runner:latest'),
      logging: LogDriver.awsLogs({
        logGroup,
        streamPrefix: 'github-actions-runner',
      }),
      environment: {
        REPOSITORY_URL: CONFIG.repositoryUrl,
        TOKEN: 'set_self-hosted_runner_token',
      },
    });

    taskDefinition.addVolume({
      name: 'docker_sock',
      host: { sourcePath: '/var/run/docker.sock' },
    });
    container.addMountPoints({
      containerPath: '/var/run/docker.sock',
      sourceVolume: 'docker_sock',
      readOnly: true,
    });
  }

  /**
   * Create a security group for cluster node
   *
   * @returns Security Group
   */
  private createSecurityGroup() {
    const securityGroup = new SecurityGroup(this, `${this.clusterName}-sg`, {
      securityGroupName: `${this.clusterName}-sg`,
      vpc: this.vpc,
      description: `SG for ${this.clusterName}`,
      allowAllOutbound: true,
    });
    securityGroup.addIngressRule(Peer.anyIpv4(), Port.tcp(22), 'Allow SSH Access');
    return securityGroup;
  }

  /**
   * Add autoscaling group capacity
   *
   * @param nodeName cluster node name
   * @param node cluster node configuration
   * @returns Autoscaling group capacity provider
   */
  private addAsgCapacity(nodeName: string, node: NodeConfig) {
    const autoScalingGroup = new AutoScalingGroup(this, `ASG-${nodeName}`, {
      vpc: this.vpc,
      autoScalingGroupName: `${this.clusterName}-${nodeName}`,
      instanceType: node.instanceType,
      machineImage: node.machineImage,
      spotPrice: node.spotPrice,
      minCapacity: node.minCapacity,
      maxCapacity: node.maxCapacity,
      keyName: node.sshKey,
      securityGroup: this.securityGroup,
      terminationPolicies: [
        TerminationPolicy.OLDEST_INSTANCE,
        TerminationPolicy.DEFAULT,
      ],
    });
    const capacityProvider = new AsgCapacityProvider(this, `AsgCapacityProvider-${nodeName}`, {
      capacityProviderName: `${nodeName}-${this.clusterName}-asg-cp`,
      spotInstanceDraining: true,
      autoScalingGroup,
      enableManagedScaling: true,
    });
    this.cluster.addAsgCapacityProvider(capacityProvider);
    return capacityProvider;
  }
}
