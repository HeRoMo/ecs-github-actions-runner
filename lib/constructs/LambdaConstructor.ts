import {
  Effect,
  IRole,
  ManagedPolicy,
  PolicyDocument,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';
import { CONFIG } from '../Config';

import { EcsGithubActionsRunnerInfo } from './EcsGithubActionsRunner';

/**
 * Lambda Function constructor to operate Github actions runner tasks.
 */
export class LambdaConstructor extends Construct {
  private readonly startRunnerFunc: NodejsFunction;
  private readonly stopRunnerFunc: NodejsFunction;

  /**
   * Constructor
   *
   * @param scope Scope
   * @param id Construct name
   * @param taskInfo Information of Github actions runner cluster
   */
  constructor(
    scope: Construct,
    private readonly id: string,
    private readonly taskInfo: EcsGithubActionsRunnerInfo,
  ) {
    super(scope, id);

    const role = this.createFunctionRole();

    this.startRunnerFunc = this.createFunction(
      `${this.id}-start-runners`,
      'lambda/handlers/startRunners.ts',
      'handler',
      role,
    );
    this.stopRunnerFunc = this.createFunction(
      `${this.id}-stop-runners`,
      'lambda/handlers/stopRunners.ts',
      'handler',
      role,
    );
  }

  /**
   * Create the role of functions.
   *
   * @returns Role.
   */
  private createFunctionRole(
    region: string = CONFIG.region || '*',
    account: string = process.env.CDK_DEFAULT_ACCOUNT || '*',
  ) {
    const path = `/${this.id}/`;
    const roleName = 'ProvisionRunnersRole';
    const role = new Role(this, `${this.id}-role`, {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      path,
      roleName,
      description: 'Role of Lambda function to start or stop Gihhub runners',
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
      ],
      inlinePolicies: {
        runTask: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'ecs:ListTaskDefinitions',
                'ecs:ListTasks',
              ],
              resources: [
                '*',
              ],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'ecs:runTask',
                'ecs:stopTask',
              ],
              resources: [
                `arn:aws:ecs:${region}:${account}:task/${this.taskInfo.clusterName}/*`,
                `arn:aws:ecs:${region}:${account}:task-definition/${this.taskInfo.ec2TaskDefFamily}:*`,
                ...this.taskInfo.fargateTaskDefFamilies.map((family) => (
                  `arn:aws:ecs:${region}:${account}:task-definition/${family}:*`
                )),
              ],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'secretsmanager:GetSecretValue',
              ],
              resources: [
                `arn:aws:secretsmanager:${region}:${account}:secret:*`,
              ],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'iam:PassRole',
              ],
              resources: [
                `arn:aws:iam::${account}:role/*`,
              ],
            }),
          ],
        }),
      },
    });
    return role;
  }

  /**
   * Create a Lambda function
   *
   * @param functionName function name
   * @param entry Entry file path
   * @param handler Handler function name
   * @param role Role of functions
   * @returns Created function.
   */
  private createFunction(functionName: string, entry: string, handler: string, role: IRole) {
    return new NodejsFunction(this, functionName, {
      functionName,
      entry,
      handler,
      role,
      runtime: Runtime.NODEJS_14_X,
      bundling: {
        minify: true,
        sourceMap: true,
        target: 'es2020',
      },
      logRetention: RetentionDays.ONE_WEEK,
      environment: {
        CLUSTER_NAME: this.taskInfo.clusterName,
        CONTAINER_NAME: this.taskInfo.containerName,
        EC2_TASK_DEF_FAMILY: `${this.taskInfo.ec2TaskDefFamily}`,
        EC2_CAPACITY_PROVIDERS: this.taskInfo.ec2CapacityProviders.join(','),
        FARGATE_TASK_DEF_FAMILIES: this.taskInfo.fargateTaskDefFamilies.join(','),
        REPO_OWNER: CONFIG.repo.owner,
        REPO_NAME: CONFIG.repo.name,
        SECRET_NAME: CONFIG.secretName,
      },
    });
  }
}
