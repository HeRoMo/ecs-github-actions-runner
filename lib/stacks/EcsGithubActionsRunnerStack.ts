import { Stack } from 'aws-cdk-lib';
import { Vpc } from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';
import { EcsGithubActionsRunner } from '../constructs/EcsGithubActionsRunner';

/**
 * Construct ECS cluster of Github actions runner.
 */
export class EcsGithubActionsRunnerStack extends Stack {
  private readonly clusterConstructor: EcsGithubActionsRunner;

  constructor(scope: Construct, clusterName: string, private readonly vpc: Vpc) {
    super(scope, `${clusterName}-stack`);

    this.clusterConstructor = new EcsGithubActionsRunner(this, clusterName, vpc);
  }

  /**
   * runnerTaskInfo
   */
  public get runnerTaskInfo() {
    return this.clusterConstructor.clusterInfo;
  }
}
