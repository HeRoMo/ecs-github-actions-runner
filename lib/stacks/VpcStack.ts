import { Stack, StackProps } from 'aws-cdk-lib';
import {
  SubnetConfiguration, SubnetType, Vpc, VpcProps,
} from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

/**
 * Construct VPC stack
 */
export class VpcStack extends Stack {
  public readonly vpc: Vpc;
  constructor(scope: Construct, vpcName: string, props?: StackProps | undefined) {
    super(scope, `${vpcName}-stack`, props)

    const subnetConfiguration: SubnetConfiguration[] = [
      { name: 'Public', cidrMask: 24, subnetType: SubnetType.PUBLIC },
    ];

    const vpcProps: VpcProps = {
      vpcName,
      cidr: '10.0.0.0/16',
      vpnGateway: false,
      natGateways: 0,
      subnetConfiguration,
    };
    this.vpc = new Vpc(this, vpcName, vpcProps);
  }
}
