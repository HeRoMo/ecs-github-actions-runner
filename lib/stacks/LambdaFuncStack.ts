import { Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { EcsGithubActionsRunnerInfo } from '../constructs/EcsGithubActionsRunner';
import { LambdaConstructor } from '../constructs/LambdaConstructor';

/**
 * Construnt Lambda functions to operate Github runner tasks.
 */
export class LambdaFuncStack extends Stack {
  constructor(scope: Construct, id: string, taskInfo: EcsGithubActionsRunnerInfo) {
    super(scope, `${id}-stack`);

    new LambdaConstructor(this, id, taskInfo);
  }
}
