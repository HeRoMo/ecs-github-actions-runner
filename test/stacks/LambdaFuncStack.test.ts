import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { EcsGithubActionsRunnerInfo } from '../../lib/constructs/EcsGithubActionsRunner';
import { LambdaFuncStack } from '../../lib/stacks/LambdaFuncStack';

test('Snapshot Test', () => {
  const app = new cdk.App();
  const info: EcsGithubActionsRunnerInfo = {
    clusterName: 'test-cluster',
    containerName: 'github-actions-runner',
    ec2TaskDefFamily: 'test-cluster-ec2',
    ec2CapacityProviders: [
      'amd64-cap-provider',
      'arm64-cap-provider',
    ],
    fargateTaskDefFamilies: [
      'test-cluster-fg-arm64',
      'test-cluster-fg-x86_64',
    ],
  };
  const funcStack = new LambdaFuncStack(app, 'test-runner', info);

  const template = Template.fromStack(funcStack);
  expect(template).toMatchSnapshot();
});
