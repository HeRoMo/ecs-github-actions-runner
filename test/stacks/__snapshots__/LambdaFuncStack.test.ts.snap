// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
Object {
  "Description": "Lambda functions to operate Tasks of test-cluster",
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-us-east-1",
          },
          "S3Key": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testrunnertestclusterstartrunnersec204C3B8B9": Object {
      "DependsOn": Array [
        "testrunnertestrunnerrole26678EE0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-us-east-1",
          },
          "S3Key": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "CLUSTER_NAME": "test-cluster",
            "CONTAINER_NAME": "github-actions-runner",
            "EC2_CAPACITY_PROVIDERS": "amd64-cap-provider,arm64-cap-provider",
            "EC2_TASK_DEF_FAMILY": "test-cluster-ec2",
            "FARGATE_TASK_DEF_FAMILIES": "test-cluster-fg-arm64,test-cluster-fg-x86_64",
            "REPO_NAME": "reponame",
            "REPO_OWNER": "your",
            "SECRET_NAME": "test-secret-name",
            "SECURITY_GROUP": "sg-XXXXXXXXXXXXXXXXX",
            "SUBNETS": "subnet-11111111111111111,subnet-22222222222222222",
          },
        },
        "FunctionName": "test-cluster-start-runners-ec2",
        "Handler": "index.ec2Handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testrunnertestrunnerrole26678EE0",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testrunnertestclusterstartrunnersec2LogRetentionD95CC791": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "testrunnertestclusterstartrunnersec204C3B8B9",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "testrunnertestclusterstartrunnersfargateDE56FBAE": Object {
      "DependsOn": Array [
        "testrunnertestrunnerrole26678EE0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-us-east-1",
          },
          "S3Key": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "CLUSTER_NAME": "test-cluster",
            "CONTAINER_NAME": "github-actions-runner",
            "EC2_CAPACITY_PROVIDERS": "amd64-cap-provider,arm64-cap-provider",
            "EC2_TASK_DEF_FAMILY": "test-cluster-ec2",
            "FARGATE_TASK_DEF_FAMILIES": "test-cluster-fg-arm64,test-cluster-fg-x86_64",
            "REPO_NAME": "reponame",
            "REPO_OWNER": "your",
            "SECRET_NAME": "test-secret-name",
            "SECURITY_GROUP": "sg-XXXXXXXXXXXXXXXXX",
            "SUBNETS": "subnet-11111111111111111,subnet-22222222222222222",
          },
        },
        "FunctionName": "test-cluster-start-runners-fargate",
        "Handler": "index.fargateHandler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testrunnertestrunnerrole26678EE0",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testrunnertestclusterstartrunnersfargateLogRetentionBD39DC4E": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "testrunnertestclusterstartrunnersfargateDE56FBAE",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "testrunnertestclusterstoprunners0DD77B77": Object {
      "DependsOn": Array [
        "testrunnertestrunnerrole26678EE0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-us-east-1",
          },
          "S3Key": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "CLUSTER_NAME": "test-cluster",
            "CONTAINER_NAME": "github-actions-runner",
            "EC2_CAPACITY_PROVIDERS": "amd64-cap-provider,arm64-cap-provider",
            "EC2_TASK_DEF_FAMILY": "test-cluster-ec2",
            "FARGATE_TASK_DEF_FAMILIES": "test-cluster-fg-arm64,test-cluster-fg-x86_64",
            "REPO_NAME": "reponame",
            "REPO_OWNER": "your",
            "SECRET_NAME": "test-secret-name",
            "SECURITY_GROUP": "sg-XXXXXXXXXXXXXXXXX",
            "SUBNETS": "subnet-11111111111111111,subnet-22222222222222222",
          },
        },
        "FunctionName": "test-cluster-stop-runners",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testrunnertestrunnerrole26678EE0",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testrunnertestclusterstoprunnersLogRetentionC77F0BB5": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "testrunnertestclusterstoprunners0DD77B77",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "testrunnertestrunnerrole26678EE0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role of Lambda function to start or stop Gihhub runners",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Path": "/test-runner/",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "ecs:ListTaskDefinitions",
                    "ecs:ListTasks",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "ecs:runTask",
                    "ecs:stopTask",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    "arn:aws:ecs:us-east-1:*:task/test-cluster/*",
                    "arn:aws:ecs:us-east-1:*:task-definition/test-cluster-ec2:*",
                    "arn:aws:ecs:us-east-1:*:task-definition/test-cluster-fg-arm64:*",
                    "arn:aws:ecs:us-east-1:*:task-definition/test-cluster-fg-x86_64:*",
                  ],
                },
                Object {
                  "Action": "secretsmanager:GetSecretValue",
                  "Effect": "Allow",
                  "Resource": "arn:aws:secretsmanager:us-east-1:*:secret:*",
                },
                Object {
                  "Action": "iam:PassRole",
                  "Effect": "Allow",
                  "Resource": "arn:aws:iam::*:role/*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "runTask",
          },
        ],
        "RoleName": "ProvisionRunnersRole",
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
